retrofit-http-logging/src/main/java/retrofit/HttpLoggingInterceptor.java

```diff
@@ -0,0 +1,234 @@
+    this.level = level;
+  }
+
+  @Override public Response intercept(Chain chain) throws IOException {
+    Level level = this.level;
+
+    Request request = chain.request();
+    if (level == Level.NONE) {
+      return chain.proceed(request);
+    }
+
+    boolean logBody = level == Level.BODY;
+    boolean logHeaders = logBody || level == Level.HEADERS;
+
+    RequestBody requestBody = request.body();
+    boolean hasRequestBody = requestBody != null;
+
+    Connection connection = chain.connection();
+    Protocol protocol = connection != null ? connection.getProtocol() : Protocol.HTTP_1_1;
+    String requestStartMessage =
+        "--> " + request.method() + ' ' + requestPath(request.httpUrl()) + ' ' + protocol(protocol);
+    if (!logHeaders && hasRequestBody) {
+      requestStartMessage += " (" + requestBody.contentLength() + "-byte body)";
+    }
+    logger.log(requestStartMessage);
+
+    if (logHeaders) {
+      Headers headers = request.headers();
+      for (int i = 0, count = headers.size(); i < count; i++) {
+        logger.log(headers.name(i) + ": " + headers.value(i));
+      }
+
+      if (logBody && hasRequestBody) {
+        Buffer buffer = new Buffer();
+        requestBody.writeTo(buffer);
+
+        Charset charset = UTF8;
+        MediaType contentType = requestBody.contentType();
+        if (contentType != null) {
+          contentType.charset(UTF8);
+        }
+
+        logger.log("");
+        logger.log(buffer.readString(charset));
+      }
+
+      String endMessage = "--> END " + request.method();
+      if (logBody && hasRequestBody) {
+        endMessage += " (" + requestBody.contentLength() + "-byte body)";
+      }
+      logger.log(endMessage);
+    }
+
+    long startNs = System.nanoTime();
+    Response response = chain.proceed(request);
+    long tookMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNs);
+
+    ResponseBody responseBody = response.body();
+    logger.log("<-- " + protocol(response.protocol()) + ' ' + response.code() + ' '
+        + response.message() + " (" + tookMs + "ms"
+        + (!logHeaders ? ", " + responseBody.contentLength() + "-byte body" : "") + ')');
+
+    if (logHeaders) {
+      Headers headers = response.headers();
+      for (int i = 0, count = headers.size(); i < count; i++) {
+        logger.log(headers.name(i) + ": " + headers.value(i));
+      }
+
+      if (logBody) {
+        Buffer buffer = new Buffer();
+        responseBody.source().readAll(buffer);
+
+        Charset charset = UTF8;
+        MediaType contentType = responseBody.contentType();
+        if (contentType != null) {
+          charset = contentType.charset(UTF8);
+        }
+
+        if (responseBody.contentLength() > 0) {
+          logger.log("");
+          logger.log(buffer.clone().readString(charset));
+        }
+
+        // Since we consumed the original, replace the one-shot body in the response with a new one.
+        response = response.newBuilder()
```

Comment:
If you wanna save some work, you could load the full body into a buffer and call clone().